File: libs/scheduling-cron/src/parser.rs
//! Cron expression parser. Accepts both the Unix 5-field and the
//! Quartz 6-field (seconds-prefixed) flavours and the special
//! characters listed in the Foundry trigger reference.

use chrono_tz::Tz;
use thiserror::Error;

use crate::schedule::{CronFlavor, CronSchedule, DayOfMonth, DayOfWeek, ValueSet};

#[derive(Debug, Error, PartialEq, Eq)]
pub enum CronError {
    #[error("cron expression must have {expected} fields, found {found}")]
    WrongFieldCount { expected: usize, found: usize },
    #[error("invalid value '{value}' in {field} field: {reason}")]
    InvalidValue {
        field: &'static str,
        value: String,
        reason: &'static str,
    },
    #[error("value {value} is out of range {min}..={max} for {field}")]
    OutOfRange {
        field: &'static str,
        value: u32,
        min: u32,
        max: u32,
    },
    #[error("step value must be greater than zero in {field}")]
    InvalidStep { field: &'static str },
    #[error("'{character}' is not allowed in the {field} field")]
    DisallowedSpecial {
        field: &'static str,
        character: char,
    },
}

/// Entry point. Parses `expr` according to `flavor`, attaching the
/// resulting schedule to `tz`.
pub fn parse_cron(expr: &str, flavor: CronFlavor, tz: Tz) -> Result<CronSchedule, CronError> {
    let trimmed = expr.trim();
    let parts: Vec<&str> = trimmed.split_whitespace().collect();
    let expected = match flavor {
        CronFlavor::Unix5 => 5,
        CronFlavor::Quartz6 => 6,
    };
    if parts.len() != expected {
        return Err(CronError::WrongFieldCount {
            expected,
            found: parts.len(),
        });
    }

    let (seconds, idx) = if matches!(flavor, CronFlavor::Quartz6) {
        (parse_value_set(parts[0], "seconds", 0, 59, &[])?, 1)
    } else {
        // Unix-5 always fires on second 0 of the matched minute.
        (ValueSet::from_values(vec![0]), 0)
    };

    let minutes = parse_value_set(parts[idx], "minutes", 0, 59, &[])?;
    let hours = parse_value_set(parts[idx + 1], "hours", 0, 23, &[])?;
    let day_of_month = parse_day_of_month(parts[idx + 2])?;
    let months = parse_month_field(parts[idx + 3])?;
    let day_of_week = parse_day_of_week(parts[idx + 4])?;

    Ok(CronSchedule {
        flavor,
        seconds,
        minutes,
        hours,
        day_of_month,
        months,
        day_of_week,
        time_zone: tz,
    })
}

/// Generic numeric value-set parser. Supports `*`, `-`, `/`, `,`.
/// `name_aliases` maps textual names (case-insensitive) to integer
/// values — used by month and day-of-week fields.
fn parse_value_set(
    raw: &str,
    field: &'static str,
    min: u32,
    max: u32,
    name_aliases: &[(&str, u32)],
) -> Result<ValueSet, CronError> {
    if raw == "*" {
        return Ok(ValueSet::star());
    }

    let mut values: Vec<u32> = Vec::new();
    for component in raw.split(',') {
        for value in expand_component(component, field, min, max, name_aliases)? {
            values.push(value);
        }
    }
    Ok(ValueSet::from_values(values))
}

fn expand_component(
    component: &str,
    field: &'static str,
    min: u32,
    max: u32,
    name_aliases: &[(&str, u32)],
) -> Result<Vec<u32>, CronError> {
    if component.is_empty() {
        return Err(CronError::InvalidValue {
            field,
            value: component.to_string(),
            reason: "empty component",
        });
    }

    let (range_part, step) = match component.split_once('/') {
        Some((range, step_raw)) => {
            let step: u32 = step_raw.parse().map_err(|_| CronError::InvalidValue {
                field,
                value: step_raw.to_string(),
                reason: "step must be a non-negative integer",
            })?;
            if step == 0 {
                return Err(CronError::InvalidStep { field });
            }
            (range, step)
        }
        None => (component, 1u32),
    };

    let (start, end) = if range_part == "*" {
        (min, max)
    } else if let Some((from, to)) = range_part.split_once('-') {
        let from = parse_int_or_alias(from, field, min, max, name_aliases)?;
        let to = parse_int_or_alias(to, field, min, max, name_aliases)?;
        (from, to)
    } else {
        let value = parse_int_or_alias(range_part, field, min, max, name_aliases)?;
        // `25/10` means "every 10th value starting from 25 up to max".
        if step != 1 {
            (value, max)
        } else {
            (value, value)
        }
    };

    if start > end {
        return Err(CronError::InvalidValue {
            field,
            value: component.to_string(),
            reason: "range start is greater than range end",
        });
    }
    if start < min || end > max {
        return Err(CronError::OutOfRange {
            field,
            value: if start < min { start } else { end },
            min,
            max,
        });
    }

    let mut out = Vec::new();
    let mut value = start;
    while value <= end {
        out.push(value);
        value = value.saturating_add(step);
        if value < start {
            break;
        }
    }
    Ok(out)
}

fn parse_int_or_alias(
    raw: &str,
    field: &'static str,
    min: u32,
    max: u32,
    name_aliases: &[(&str, u32)],
) -> Result<u32, CronError> {
    let raw_trim = raw.trim();
    if raw_trim.is_empty() {
        return Err(CronError::InvalidValue {
            field,
            value: raw.to_string(),
            reason: "empty integer",
        });
    }
    if let Some(alias_value) = name_aliases
        .iter()
        .find(|(name, _)| name.eq_ignore_ascii_case(raw_trim))
        .map(|(_, value)| *value)
    {
        return Ok(alias_value);
    }
    let value: u32 = raw_trim.parse().map_err(|_| CronError::InvalidValue {
        field,
        value: raw.to_string(),
        reason: "not a valid integer or alias",
    })?;
    if value < min || value > max {
        return Err(CronError::OutOfRange {
            field,
            value,
            min,
            max,
        });
    }
    Ok(value)
}

fn parse_month_field(raw: &str) -> Result<ValueSet, CronError> {
    static ALIASES: &[(&str, u32)] = &[
        ("JAN", 1),
        ("FEB", 2),
        ("MAR", 3),
        ("APR", 4),
        ("MAY", 5),
        ("JUN", 6),
        ("JUL", 7),
        ("AUG", 8),
        ("SEP", 9),
        ("OCT", 10),
        ("NOV", 11),
        ("DEC", 12),
    ];
    parse_value_set(raw, "month", 1, 12, ALIASES)
}

fn day_of_week_aliases() -> &'static [(&'static str, u32)] {
    &[
        ("SUN", 0),
        ("MON", 1),
        ("TUE", 2),
        ("WED", 3),
        ("THU", 4),
        ("FRI", 5),
        ("SAT", 6),
    ]
}

fn parse_day_of_month(raw: &str) -> Result<DayOfMonth, CronError> {
    if raw == "*" {
        return Ok(DayOfMonth::Star);
    }
    if raw.eq_ignore_ascii_case("L") {
        return Ok(DayOfMonth::Last);
    }
    if raw.contains('#') || raw.to_ascii_uppercase().contains('L') {
        return Err(CronError::DisallowedSpecial {
            field: "day-of-month",
            character: '#',
        });
    }
    let set = parse_value_set(raw, "day-of-month", 1, 31, &[])?;
    Ok(DayOfMonth::Set(set))
}

fn parse_day_of_week(raw: &str) -> Result<DayOfWeek, CronError> {
    if raw == "*" {
        return Ok(DayOfWeek::Star);
    }
    let aliases = day_of_week_aliases();

    // Bare `L` in DOW means "Saturday" (the last day of the week)
    // recurring every week — not "last Saturday of the month".
    // Per Foundry trigger reference table.
    if raw.eq_ignore_ascii_case("L") {
        return Ok(DayOfWeek::Set(ValueSet::from_values(vec![6])));
    }

    // `<weekday>L` form, e.g. `2L`, `MONL`.
    if let Some(prefix) = raw
        .strip_suffix('L')
        .or_else(|| raw.strip_suffix('l'))
    {
        let weekday = parse_int_or_alias(prefix, "day-of-week", 0, 7, aliases)?;
        let weekday = normalise_weekday(weekday);
        return Ok(DayOfWeek::LastWeekdayOfMonth { weekday });
    }

    // `<weekday>#<n>` form.
    if let Some((weekday_raw, occurrence_raw)) = raw.split_once('#') {
        let weekday = parse_int_or_alias(weekday_raw, "day-of-week", 0, 7, aliases)?;
        let weekday = normalise_weekday(weekday);
        let occurrence: u32 =
            occurrence_raw
                .trim()
                .parse()
                .map_err(|_| CronError::InvalidValue {
                    field: "day-of-week",
                    value: occurrence_raw.to_string(),
                    reason: "occurrence after '#' must be an integer",
                })?;
        if !(1..=5).contains(&occurrence) {
            return Err(CronError::OutOfRange {
                field: "day-of-week-occurrence",
                value: occurrence,
                min: 1,
                max: 5,
            });
        }
        return Ok(DayOfWeek::NthWeekdayOfMonth {
            weekday,
            occurrence,
        });
    }

    let raw_set = parse_value_set(raw, "day-of-week", 0, 7, aliases)?;
    let normalised = normalise_weekday_set(raw_set);
    Ok(DayOfWeek::Set(normalised))
}

fn normalise_weekday(value: u32) -> u32 {
    if value == 7 {
        0
    } else {
        value
    }
}

fn normalise_weekday_set(set: ValueSet) -> ValueSet {
    if set.is_star() {
        return set;
    }
    let mut values: Vec<u32> = set.values.into_iter().map(normalise_weekday).collect();
    values.sort_unstable();
    values.dedup();
    ValueSet {
        values,
        star: false,
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use chrono_tz::UTC;

    #[test]
    fn parses_simple_unix_five_field() {
        let s = parse_cron("30 9 * * 1", CronFlavor::Unix5, UTC).expect("parse");
        assert!(s.minutes.contains(30));
        assert!(s.hours.contains(9));
        assert!(s.months.is_star());
        assert_eq!(s.seconds.values, vec![0]);
        match s.day_of_week {
            DayOfWeek::Set(set) => assert_eq!(set.values, vec![1]),
            other => panic!("expected weekday set, got {other:?}"),
        }
    }

    #[test]
    fn parses_quartz_seconds_prefix() {
        let s = parse_cron("0 30 9 * * 1", CronFlavor::Quartz6, UTC).expect("parse");
        assert!(s.seconds.contains(0));
        assert!(s.minutes.contains(30));
    }

    #[test]
    fn rejects_too_many_fields() {
        let err = parse_cron("0 0 * * * 1", CronFlavor::Unix5, UTC).unwrap_err();
        assert!(matches!(err, CronError::WrongFieldCount { .. }));
    }

    #[test]
    fn parses_step_with_value_starting_point() {
        let s = parse_cron("25/10 * * * *", CronFlavor::Unix5, UTC).expect("parse");
        assert_eq!(s.minutes.values, vec![25, 35, 45, 55]);
    }

    #[test]
    fn parses_step_in_range() {
        let s = parse_cron("25-45/10 * * * *", CronFlavor::Unix5, UTC).expect("parse");
        assert_eq!(s.minutes.values, vec![25, 35, 45]);
    }

    #[test]
    fn parses_l_in_dom_field() {
        let s = parse_cron("0 9 L * *", CronFlavor::Unix5, UTC).expect("parse");
        assert!(matches!(s.day_of_month, DayOfMonth::Last));
    }

    #[test]
    fn parses_2l_in_dow_field() {
        let s = parse_cron("0 9 * * 2L", CronFlavor::Unix5, UTC).expect("parse");
        match s.day_of_week {
            DayOfWeek::LastWeekdayOfMonth { weekday } => assert_eq!(weekday, 2),
            other => panic!("expected last-weekday, got {other:?}"),
        }
    }

    #[test]
    fn parses_3hash1_in_dow_field() {
        let s = parse_cron("0 9 * 4 3#1", CronFlavor::Unix5, UTC).expect("parse");
        match s.day_of_week {
            DayOfWeek::NthWeekdayOfMonth {
                weekday,
                occurrence,
            } => {
                assert_eq!(weekday, 3);
                assert_eq!(occurrence, 1);
            }
            other => panic!("expected nth-weekday, got {other:?}"),
        }
    }

    #[test]
    fn accepts_month_name_alias() {
        let s = parse_cron("0 9 1 MAR *", CronFlavor::Unix5, UTC).expect("parse");
        assert_eq!(s.months.values, vec![3]);
    }

    #[test]
    fn weekday_seven_normalises_to_zero() {
        let s = parse_cron("0 0 * * 7", CronFlavor::Unix5, UTC).expect("parse");
        match s.day_of_week {
            DayOfWeek::Set(set) => assert_eq!(set.values, vec![0]),
            other => panic!("expected weekday set, got {other:?}"),
        }
    }
}

File: libs/scheduling-cron/parser.go
package schedulingcron

import (
	"fmt"
	"strconv"
	"strings"
	"time"
)

// CronError is the typed error tree returned by ParseCron. Use
// errors.As / the Is* helpers to discriminate.
type CronError struct {
	Kind     CronErrorKind
	Field    string
	Value    string
	Reason   string
	IntValue uint32
	Min      uint32
	Max      uint32
	Found    int
	Expected int
	Char     rune
}

// CronErrorKind discriminates CronError variants.
type CronErrorKind uint8

const (
	// ErrWrongFieldCount — expression had wrong number of
	// whitespace-separated fields for the chosen flavour.
	ErrWrongFieldCount CronErrorKind = iota + 1
	// ErrInvalidValue — a sub-component wasn't parseable.
	ErrInvalidValue
	// ErrOutOfRange — a value fell outside the field's accepted range.
	ErrOutOfRange
	// ErrInvalidStep — `/0` or non-integer step.
	ErrInvalidStep
	// ErrDisallowedSpecial — `#` in DOM, etc.
	ErrDisallowedSpecial
)

func (e *CronError) Error() string {
	switch e.Kind {
	case ErrWrongFieldCount:
		return fmt.Sprintf("cron expression must have %d fields, found %d", e.Expected, e.Found)
	case ErrInvalidValue:
		return fmt.Sprintf("invalid value '%s' in %s field: %s", e.Value, e.Field, e.Reason)
	case ErrOutOfRange:
		return fmt.Sprintf("value %d is out of range %d..=%d for %s", e.IntValue, e.Min, e.Max, e.Field)
	case ErrInvalidStep:
		return fmt.Sprintf("step value must be greater than zero in %s", e.Field)
	case ErrDisallowedSpecial:
		return fmt.Sprintf("'%c' is not allowed in the %s field", e.Char, e.Field)
	default:
		return "cron parse error"
	}
}

// ParseCron is the entry point. Parses `expr` according to `flavor`,
// attaching the resulting schedule to `tz`. Mirrors fn parse_cron.
func ParseCron(expr string, flavor CronFlavor, tz *time.Location) (CronSchedule, error) {
	trimmed := strings.TrimSpace(expr)
	parts := strings.Fields(trimmed)
	expected := 5
	if flavor == Quartz6 {
		expected = 6
	}
	if len(parts) != expected {
		return CronSchedule{}, &CronError{
			Kind: ErrWrongFieldCount, Expected: expected, Found: len(parts),
		}
	}

	var (
		seconds ValueSet
		idx     int
	)
	if flavor == Quartz6 {
		s, err := parseValueSet(parts[0], "seconds", 0, 59, nil)
		if err != nil {
			return CronSchedule{}, err
		}
		seconds = s
		idx = 1
	} else {
		// Unix-5 always fires on second 0 of the matched minute.
		seconds = FromValues([]uint32{0})
	}

	minutes, err := parseValueSet(parts[idx], "minutes", 0, 59, nil)
	if err != nil {
		return CronSchedule{}, err
	}
	hours, err := parseValueSet(parts[idx+1], "hours", 0, 23, nil)
	if err != nil {
		return CronSchedule{}, err
	}
	dom, err := parseDayOfMonth(parts[idx+2])
	if err != nil {
		return CronSchedule{}, err
	}
	months, err := parseMonthField(parts[idx+3])
	if err != nil {
		return CronSchedule{}, err
	}
	dow, err := parseDayOfWeek(parts[idx+4])
	if err != nil {
		return CronSchedule{}, err
	}

	if tz == nil {
		tz = time.UTC
	}
	return CronSchedule{
		Flavor:     flavor,
		Seconds:    seconds,
		Minutes:    minutes,
		Hours:      hours,
		DayOfMonth: dom,
		Months:     months,
		DayOfWeek:  dow,
		TimeZone:   tz,
	}, nil
}

type alias struct {
	name  string
	value uint32
}

var monthAliases = []alias{
	{"JAN", 1}, {"FEB", 2}, {"MAR", 3}, {"APR", 4}, {"MAY", 5}, {"JUN", 6},
	{"JUL", 7}, {"AUG", 8}, {"SEP", 9}, {"OCT", 10}, {"NOV", 11}, {"DEC", 12},
}

var dayOfWeekAliases = []alias{
	{"SUN", 0}, {"MON", 1}, {"TUE", 2}, {"WED", 3}, {"THU", 4}, {"FRI", 5}, {"SAT", 6},
}

// parseValueSet is the generic numeric value-set parser. Supports
// `*`, `-`, `/`, `,`. Mirrors fn parse_value_set.
func parseValueSet(raw, field string, min, max uint32, aliases []alias) (ValueSet, error) {
	if raw == "*" {
		return Star(), nil
	}
	values := []uint32{}
	for _, component := range strings.Split(raw, ",") {
		expanded, err := expandComponent(component, field, min, max, aliases)
		if err != nil {
			return ValueSet{}, err
		}
		values = append(values, expanded...)
	}
	return FromValues(values), nil
}

func expandComponent(component, field string, min, max uint32, aliases []alias) ([]uint32, error) {
	if component == "" {
		return nil, &CronError{
			Kind: ErrInvalidValue, Field: field, Value: component, Reason: "empty component",
		}
	}

	rangePart := component
	step := uint32(1)
	if i := strings.Index(component, "/"); i >= 0 {
		rangePart = component[:i]
		stepRaw := component[i+1:]
		stepParsed, err := strconv.ParseUint(stepRaw, 10, 32)
		if err != nil {
			return nil, &CronError{
				Kind: ErrInvalidValue, Field: field, Value: stepRaw,
				Reason: "step must be a non-negative integer",
			}
		}
		if stepParsed == 0 {
			return nil, &CronError{Kind: ErrInvalidStep, Field: field}
		}
		step = uint32(stepParsed)
	}

	var start, end uint32
	switch {
	case rangePart == "*":
		start, end = min, max
	case strings.Contains(rangePart, "-"):
		dash := strings.Index(rangePart, "-")
		from := rangePart[:dash]
		to := rangePart[dash+1:]
		fv, err := parseIntOrAlias(from, field, min, max, aliases)
		if err != nil {
			return nil, err
		}
		tv, err := parseIntOrAlias(to, field, min, max, aliases)
		if err != nil {
			return nil, err
		}
		start, end = fv, tv
	default:
		v, err := parseIntOrAlias(rangePart, field, min, max, aliases)
		if err != nil {
			return nil, err
		}
		// `25/10` → start=25, end=max.
		if step != 1 {
			start, end = v, max
		} else {
			start, end = v, v
		}
	}

	if start > end {
		return nil, &CronError{
			Kind: ErrInvalidValue, Field: field, Value: component,
			Reason: "range start is greater than range end",
		}
	}
	if start < min || end > max {
		bad := start
		if start >= min {
			bad = end
		}
		return nil, &CronError{
			Kind: ErrOutOfRange, Field: field, IntValue: bad, Min: min, Max: max,
		}
	}

	out := []uint32{}
	for value := start; value <= end; value += step {
		out = append(out, value)
		// Saturating add — stop on overflow.
		if value+step < value {
			break
		}
	}
	return out, nil
}

func parseIntOrAlias(raw, field string, min, max uint32, aliases []alias) (uint32, error) {
	rawTrim := strings.TrimSpace(raw)
	if rawTrim == "" {
		return 0, &CronError{
			Kind: ErrInvalidValue, Field: field, Value: raw, Reason: "empty integer",
		}
	}
	for _, a := range aliases {
		if strings.EqualFold(a.name, rawTrim) {
			return a.value, nil
		}
	}
	v, err := strconv.ParseUint(rawTrim, 10, 32)
	if err != nil {
		return 0, &CronError{
			Kind: ErrInvalidValue, Field: field, Value: raw,
			Reason: "not a valid integer or alias",
		}
	}
	if uint32(v) < min || uint32(v) > max {
		return 0, &CronError{
			Kind: ErrOutOfRange, Field: field, IntValue: uint32(v), Min: min, Max: max,
		}
	}
	return uint32(v), nil
}

func parseMonthField(raw string) (ValueSet, error) {
	return parseValueSet(raw, "month", 1, 12, monthAliases)
}

func parseDayOfMonth(raw string) (DayOfMonth, error) {
	if raw == "*" {
		return DayOfMonth{Kind: DOMStar}, nil
	}
	if strings.EqualFold(raw, "L") {
		return DayOfMonth{Kind: DOMLast}, nil
	}
	if strings.ContainsRune(raw, '#') || strings.ContainsAny(strings.ToUpper(raw), "L") {
		return DayOfMonth{}, &CronError{
			Kind: ErrDisallowedSpecial, Field: "day-of-month", Char: '#',
		}
	}
	set, err := parseValueSet(raw, "day-of-month", 1, 31, nil)
	if err != nil {
		return DayOfMonth{}, err
	}
	return DayOfMonth{Kind: DOMSet, Set: set}, nil
}

func parseDayOfWeek(raw string) (DayOfWeek, error) {
	if raw == "*" {
		return DayOfWeek{Kind: DOWStar}, nil
	}
	// Bare `L` means Saturday recurring weekly (per Foundry docs).
	if strings.EqualFold(raw, "L") {
		return DayOfWeek{Kind: DOWSet, Set: FromValues([]uint32{6})}, nil
	}
	// `<weekday>L` form.
	if strings.HasSuffix(strings.ToUpper(raw), "L") {
		prefix := raw[:len(raw)-1]
		v, err := parseIntOrAlias(prefix, "day-of-week", 0, 7, dayOfWeekAliases)
		if err != nil {
			return DayOfWeek{}, err
		}
		return DayOfWeek{Kind: DOWLastWeekdayOfMonth, Weekday: normaliseWeekday(v)}, nil
	}
	// `<weekday>#<occurrence>` form.
	if i := strings.Index(raw, "#"); i >= 0 {
		weekdayRaw := raw[:i]
		occurrenceRaw := raw[i+1:]
		weekday, err := parseIntOrAlias(weekdayRaw, "day-of-week", 0, 7, dayOfWeekAliases)
		if err != nil {
			return DayOfWeek{}, err
		}
		occurrence, err := strconv.ParseUint(strings.TrimSpace(occurrenceRaw), 10, 32)
		if err != nil {
			return DayOfWeek{}, &CronError{
				Kind: ErrInvalidValue, Field: "day-of-week", Value: occurrenceRaw,
				Reason: "occurrence after '#' must be an integer",
			}
		}
		if occurrence < 1 || occurrence > 5 {
			return DayOfWeek{}, &CronError{
				Kind: ErrOutOfRange, Field: "day-of-week-occurrence",
				IntValue: uint32(occurrence), Min: 1, Max: 5,
			}
		}
		return DayOfWeek{
			Kind:       DOWNthWeekdayOfMonth,
			Weekday:    normaliseWeekday(weekday),
			Occurrence: uint32(occurrence),
		}, nil
	}

	rawSet, err := parseValueSet(raw, "day-of-week", 0, 7, dayOfWeekAliases)
	if err != nil {
		return DayOfWeek{}, err
	}
	return DayOfWeek{Kind: DOWSet, Set: normaliseWeekdaySet(rawSet)}, nil
}

func normaliseWeekday(v uint32) uint32 {
	if v == 7 {
		return 0
	}
	return v
}

func normaliseWeekdaySet(set ValueSet) ValueSet {
	if set.IsStar() {
		return set
	}
	values := make([]uint32, 0, len(set.values))
	for _, v := range set.values {
		values = append(values, normaliseWeekday(v))
	}
	return FromValues(values)
}
